ui/apps/platform/src/hooks/useRestMutation.ts

```diff
@@ -0,0 +1,107 @@
+import { useState } from 'react';
+
+export type RequestStatus = 'idle' | 'loading' | 'success' | 'error';
+
+export type MutationOptions<Result> = {
+    /** A function to call when the mutation is successful */
+    onSuccess?: (data: Result) => void;
+    /** A function to call when the mutation fails */
+    onError?: (error: unknown) => void;
+    /** A function to call after the mutation is successful or fails */
+    onSettled?: (data: Result | undefined, error: unknown) => void;
+};
+
+export type UseRestMutationReturn<Payload, Result> = {
+    /** The result of the mutation. */
+    data: Result | undefined;
+    /** Whether the mutation is currently idle, i.e. before the mutation is first triggered */
+    isIdle: boolean;
+    /** Whether the mutation is currently loading */
+    isLoading: boolean;
+    /** Whether the mutation was successful */
+    isSuccess: boolean;
+    /** Whether the mutation failed */
+    isError: boolean;
+    /** The current status of the mutation lifecycle */
+    status: RequestStatus;
+    /** The error that occurred during the mutation */
+    error: unknown;
+    /**
+     * A function to trigger the mutation. Accepts an optional object to run callbacks for
+     * this specific mutation.
+     */
+    mutate: (payload: Payload, localOptions?: MutationOptions<Result>) => void;
+    /** A function to reset the mutation to its initial state and clear any data or error */
+    reset: () => void;
+};
+
+/**
+ * A hook used for general purpose REST mutations. The API is a subset of the one provided by
+ * `react-query`'s `useMutation` hook.
+ *
+ * @param requestFn The function that will be called when the mutation is triggered.
+ * @param options Options for the mutation to run callbacks globally each time the mutation is triggered. These will
+ *                be called before the local callbacks passed to the `mutate` function.
+ * @returns An object containing the current status of the mutation, the result of the mutation,
+ *          and a function to trigger the mutation.
+ */
+export default function useRestMutation<Payload, Result>(
+    requestFn: (payload: Payload) => Promise<Result>,
+    options: MutationOptions<Result> = {}
+): UseRestMutationReturn<Payload, Result> {
+    const [data, setData] = useState<Result>();
+    const [status, setStatus] = useState<RequestStatus>('idle');
+    const [error, setError] = useState<unknown>();
```

Comment:
Since these states have some relationship with each other and are usually set at the same time, would it be reasonable to keep the state in one object?

Reply:
I did try combining these with a `useReducer`, but it actually ended up being double the lines of code and wasn't even more readable.

I can try what you are suggesting though and use a single object since all of the states tend to update at the same time.

Reply:
👍  I like this better, I'll push an update.